storage/innobase/log/log0log.cc

```diff
@@ -3257,6 +3258,11 @@ logs_empty_and_mark_files_at_shutdown(void)
 
 			count = 0;
 		}
+		if (total_trx < total_trx_ref)
+		{
+			my_sd_extend_timeout(INNODB_EXTEND_TIMEOUT_INTERVAL);
+			total_trx_ref = total_trx;
+		}
```

Comment:
Nitpick: In the InnoDB formatting style, there must not be a line break between the `)` and `{` of an `if` statement.

The transactions here could be user transactions that are being killed. I think that this could spam `sd_notifyf()` way too often.

I think that this reporting really belongs to `trx_roll_must_shutdown()` inside the condition `recv_sys->report()`.

I believe that the rollback of a large transactions which was aborted by disconnect at shutdown could take some time. The proper place to report it would be inside `row_undo_step()`.

In fact, we could modify `trx_roll_must_shudown()` and its sole caller so that it will abort any transaction rollback on a normal (non-slow) shutdown. In that way, shutdown would  become a non-issue. The rollback would then be performed on the subsequent startup, in the background. This change would involve something like this:
```diff
diff --git a/storage/innobase/row/row0undo.cc b/storage/innobase/row/row0undo.cc
index 552b99ab4d4..c314a10842e 100644
--- a/storage/innobase/row/row0undo.cc
+++ b/storage/innobase/row/row0undo.cc
@@ -349,8 +349,7 @@ row_undo_step(
 
 	ut_ad(que_node_get_type(node) == QUE_NODE_UNDO);
 
-	if (UNIV_UNLIKELY(trx == trx_roll_crash_recv_trx)
-	    && trx_roll_must_shutdown()) {
+	if (UNIV_UNLIKELY(trx_roll_must_shutdown(trx)) {
 		/* Shutdown has been initiated. */
 		trx->error_state = DB_INTERRUPTED;
 		return(NULL);
```